<p align="center">
  <img src="header.png" width="100%">
</p>


<div style="text-align: center;">
    <strong style="display: block; margin-bottom: 10px;">Group P</strong> 
    <table style="margin: 0 auto; border-collapse: collapse; border: 1px solid black;">
        <tr>
            <th style="border: 1px solid white; padding: 8px;">Name</th>
            <th style="border: 1px solid white; padding: 8px;">Student ID</th>
        </tr>
        <tr>
            <td style="border: 1px solid white; padding: 8px;">Beatriz Monteiro</td>
            <td style="border: 1px solid white; padding: 8px;">20240591</td>
        </tr>
        <tr>
            <td style="border: 1px solid white; padding: 8px;">Catarina Nunes</td>
            <td style="border: 1px solid white; padding: 8px;">20230083</td>
        </tr>
        <tr>
            <td style="border: 1px solid white; padding: 8px;">Margarida Raposo</td>
            <td style="border: 1px solid white; padding: 8px;">20241020</td>
        </tr>
        <tr>
            <td style="border: 1px solid white; padding: 8px;">Teresa Menezes</td>
            <td style="border: 1px solid white; padding: 8px;">20240333</td>
        </tr>
    </table>
</div>

### 🔗 Table of Contents <a id='table-of-contents'></a>
1. [Introduction](#introduction)  
2. [Business Understanding](#business-understanding)  
3. [Data Understanding](#data-understanding)  
4. [Data Preparation](#data-preparation)  
5. [Modeling](#modeling)  
6. [Evaluation](#evaluation)  
7. [Conclusion](#conclusion)  

---

### <span style="background-color:#000027; padding:5px; border-radius:5px;"> 📌 Introduction <a id='introduction'></a>

This project follows the **CRISP-DM** methodology to conduct a monthly sales forecast of the smart infrastructure business unit of Siemens. 


#### Montlhy Sales Forecast

<p style="margin-bottom: 50px;"> This case study focuses on ... </p>

**Market Data** 
| Features                                      | Feature Description |
|-----------------------------------------------|---------------------|
| *MAB_ELE_PRO156*                              |  |
| *MAB_ELE_SHP156*                              |  |
| *MAB_ELE_PRO250*                              |  |
| *MAB_ELE_SHP250*                              |  |
| *MAB_ELE_PRO276*                              |  |
| *MAB_ELE_SHP276*                              |  |
| *MAB_ELE_PRO380*                              |  |
| *MAB_ELE_SHP380*                              |  |
| *MAB_ELE_PRO392*                              |  |
| *MAB_ELE_SHP392*                              |  |
| *MAB_ELE_PRO756*                              |  |
| *MAB_ELE_SHP756*                              |  |
| *MAB_ELE_PRO826*                              |  |
| *MAB_ELE_SHP826*                              |  |
| *MAB_ELE_PRO840*                              |  |
| *MAB_ELE_SHP840*                              |  |
| *MAB_ELE_PRO1100*                             |  |
| *MAB_ELE_SHP1100*                             |  |
| *RohiBASEMET1000_org*                         |  |
| *RohiENERGY1000_org*                          |  |
| *RohiMETMIN1000_org*                          |  |
| *RohiNATGAS1000_org*                          |  |
| *RohCRUDE_PETRO1000_org*                      |  |
| *RohCOPPER1000_org*                           |  |
| *WKLWEUR840_org*                              |  |
| *PRI27840_org*                                |  |
| *PRI27826_org*                                |  |
| *PRI27380_org*                                |  |
| *PRI27250_org*                                |  |
| *PRI27276_org*                                |  |
| *PRI27156_org*                                |  |
| *PRO28840_org*                                |  |
| *PRO281000_org*                               |  |
| *PRO28756_org*                                |  |
| *PRO28826_org*                                |  |
| *PRO28380_org*                                |  |
| *PRO28392_org*                                |  |
| *PRO28250_org*                                |  |
| *PRO28276_org*                                |  |
| *PRO27840_org*                                |  |
| *PRO271000_org*                               |  |
| *PRO27756_org*                                |  |
| *PRO27826_org*                                |  |
| *PRO27380_org*                                |  |
| *PRO27392_org*                                |  |
| *PRO27250_org*                                |  |
| *PRO27276_org*                                |  |

- **Production index** is an economic indicator that measures the output quantity of a country's industrial sector. In this case the base year is 2010, meaning **if the production index >100, then production decreased, compared to 2010**. 


**Sales Data**
| Features                                      | Feature Description |
|-----------------------------------------------|---------------------|
| *DATE*                                       |   |
| *Mapped_GCK*                                 |   |
| *Sales_EUR*                                  |   |

<b style="background-color:#A9A9A9; padding:5px; border-radius:5px; display: inline-block; margin-top: 50px;">CRISP-DM</b>

<ul style="margin-bottom: 30px;">
    <li><u>Business Understanding</u>: Defining objectives, assessing resources, and project planning.</li>
    <li><u>Data Understanding</u>: Collecting, exploring, and verifying data quality.</li>
    <li><u>Data Preparation</u>: Selecting, cleaning, constructing, integrating, and formatting data to ensure it is ready for analysis.</li>
    <li><u>Modeling</u>: Selecting and applying various modeling techniques while calibrating their parameters to optimal values.</li>
    <li><u>Evaluation</u>: Select the models which are the best performers and evaluate thoroughly if they align with the business objectives. </li>
    <li><u>Deployment</u>: Bridge between data mining goals and the business application of the finalized model.</li>
</ul>

<hr style="margin-top: 30px;">


In [607]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
from sklearn.preprocessing import MinMaxScaler
from plotly.subplots import make_subplots
from scipy.stats import linregress

In [608]:
pd.set_option('display.max_columns', None)

In [609]:
#compose a pallete to use in the vizualizations
pal_novaims = ['#003B5C','#003B5C','#003B5C','#003B5C','#003B5C']
pastel_color = sns.utils.set_hls_values(pal_novaims[1], l=0.4, s=0.3)

### <span style="background-color:#235987; padding:5px; border-radius:5px;"> 📌 Business Understanding <a id='business-understanding'></a>

##### Click [here](#table-of-contents) ⬆️ to return to the Index.
---

The **Business Understanding** phase of the project entails the comprehension of the background leading to the project, as well as the business goals and requirements to be achieved. 

<b style="background-color:#A9A9A9; padding:5px; border-radius:5px;">Primary Business Objective</b> : 


<b style="background-color:#A9A9A9; padding:5px; border-radius:5px;">Plan</b> : 

### <span style="background-color:#235987; padding:5px; border-radius:5px;"> 📌 Data Understanding</span> <a id='data-understanding'></a>

- **[Data Loading and Description](#data-loading-and-description)**  
- **[Data Types](#Data-TypesDU)**
- **[Univariate EDA: Descriptive Summary](#Descriptive-Summary)**
- **[Univariate EDA: Missing values](#missing-valuesDU)**  
- **[Inconsistencies](#inconsistenciesDU)**  
- **[Feature Engineering](#feature-engineeringDU)**  
- **[Univariate EDA: Data Visualization](#univariate-vizualization)**  
    - **Numerical Variables:**  
        - [Numeric variables: Histograms](#hist)
        - [Outliers Analysis: Box-Plots](#box)
    - **Categorical Variables**  
        - [Categorical variables: Bar Plots](#bar)
        - [Categorical variables: Geographic Map](#GeographicMap)
- **[Bivariate EDA: Data Visualization](#Bivariate-Vizualization)**  
   - [Numeric-Numeric: Correlations](#NNCorrelations)
   - [Numeric-Categorical: Correlations](#NCCorrelations)
   - [Categorical-Categorical: Cross-tabulations](#CCCross-tabulations)
- **[Multivariate EDA: Duplicates](#Multivariate)**
   - [Old Segmentation Vs. All](#old-segmentation)
   - [Duplicates](#duplicatesdu)  
- **[Market Basket Analysis](#MBA)**


##### Click [here](#table-of-contents) ⬆️ to return to the Index.
---

#### <span style="background-color:#235987; padding:5px; border-radius:5px;">**Data Loading and Description**</span> <a id='data-loading-and-description'></a>  
_This section provides an overview of the dataset, including its structure, size, and general characteristics._  

##### Click [here](#table-of-contents) ⬆️ to return to the Index.


In [610]:
df_market = pd.read_excel('Data/Case2_Market data.xlsx', sheet_name="Original Values", index_col=0, header=2)
df_market.head()

MAB_ELE_PRO156  MAB_ELE_SHP156  MAB_ELE_PRO250  MAB_ELE_SHP250  \
date                                                                      
 2004m2         16.9407         16.9407        112.0913         83.4589   
 2004m3         23.7119         23.7119        136.3280        106.1682   
 2004m4         24.4352         24.4352        117.7918         92.0076   
 2004m5         23.7081         23.7081        109.0025         85.6965   
 2004m6         27.0091         27.0091        133.7857        106.6415   

         MAB_ELE_PRO276  MAB_ELE_SHP276  MAB_ELE_PRO380  MAB_ELE_SHP380  \
date                                                                      
 2004m2         82.6230         79.4525        124.2896         86.5605   
 2004m3        100.5566         97.0129        143.4117        106.3445   
 2004m4         89.6532         84.9324        129.0838         95.5797   
 2004m5         86.8806         82.3728        135.5904        100.0870   
 2004m6         99.0108         95.1087        136.4249        110.8897   

         MAB_ELE_PRO392  MAB_ELE_SHP392  MAB_ELE_PRO756  MAB_ELE_SHP756  \
date                                                                      
 2004m2        109.3340        110.4953         91.2219         89.9873   
 2004m3        140.8846        144.6862         85.8663         79.8836   
 2004m4        105.8536        102.6558         85.6225         79.7408   
 2004m5        101.8648        100.3053         85.3787         79.5980   
 2004m6        120.3329        119.6164         85.1349         79.4552   

         MAB_ELE_PRO826  MAB_ELE_SHP826  MAB_ELE_PRO840  MAB_ELE_SHP840  \
date                                                                      
 2004m2        111.3538         73.6013        107.6014         79.2402   
 2004m3        127.5586         84.0476        110.1874         98.6190   
 2004m4        108.7323         73.0260        108.1666         89.7740   
 2004m5        110.6452         74.5919        108.4259         87.4638   
 2004m6        122.0210         82.3433        110.5699         97.3645   

         MAB_ELE_PRO1100  MAB_ELE_SHP1100  RohiBASEMET1000_org  \
date                                                             
 2004m2          97.1229          80.0985              54.0398   
 2004m3         113.7839          96.0159              54.6662   
 2004m4         101.7152          85.1672              54.8727   
 2004m5         101.2757          84.4858              51.2304   
 2004m6         112.0572          96.9633              52.8763   

         RohiENERGY1000_org  RohiMETMIN1000_org  RohiNATGAS1000_org  \
date                                                                  
 2004m2             44.1233             48.7479             87.0770   
 2004m3             47.5890             49.2562             87.1927   
 2004m4             47.7790             49.4238             91.3799   
 2004m5             53.5909             46.4684             99.0445   
 2004m6             50.7996             47.8039             98.6363   

         RohCRUDE_PETRO1000_org  RohCOPPER1000_org  WKLWEUR840_org  \
date                                                                 
 2004m2                 39.6395            36.6238          1.2646   
 2004m3                 42.5920            39.9311          1.2262   
 2004m4                 42.6506            39.1349          1.1985   
 2004m5                 47.5171            36.2784          1.2007   
 2004m6                 44.9676            35.6574          1.2138   

         PRI27840_org  PRI27826_org  PRI27380_org  PRI27250_org  PRI27276_org  \
date                                                                            
 2004m2       78.9699       80.7574       93.0200           NaN       93.2305   
 2004m3       79.6736       80.9621       93.5403           NaN       93.3357   
 2004m4       80.3376       80.7574       93.8524           NaN       93.4409   
 2004m5       80.7988       80.7574       93.852

In [611]:
df_market.tail()

MAB_ELE_PRO156  MAB_ELE_SHP156  MAB_ELE_PRO250  MAB_ELE_SHP250  \
date                                                                      
2021m12        310.7632        310.7632        100.5657        134.5895   
 2022m1        235.9561        235.9561         85.7435        108.1563   
 2022m2        235.9561        235.9561         90.6035        117.7158   
 2022m3        329.4134        329.4134        107.8435        136.8587   
 2022m4        267.3731        267.3731         87.6981        116.5287   

         MAB_ELE_PRO276  MAB_ELE_SHP276  MAB_ELE_PRO380  MAB_ELE_SHP380  \
date                                                                      
2021m12        118.1033        149.3643         94.0068        150.4827   
 2022m1         94.5506        120.3534         86.8510        101.2583   
 2022m2        103.9879        129.3837        106.5838        120.9565   
 2022m3        121.3081        151.2013        124.6380        153.6451   
 2022m4         99.5222        127.0229        103.5567        128.7333   

         MAB_ELE_PRO392  MAB_ELE_SHP392  MAB_ELE_PRO756  MAB_ELE_SHP756  \
date                                                                      
2021m12        127.7717        131.0297        106.7040        104.8192   
 2022m1        110.4602        110.8235        103.4993        101.7016   
 2022m2        117.8796        118.3002        100.2945         98.5840   
 2022m3        152.0006        156.4006         97.0897         95.4663   
 2022m4        114.2623        115.0120             NaN             NaN   

         MAB_ELE_PRO826  MAB_ELE_SHP826  MAB_ELE_PRO840  MAB_ELE_SHP840  \
date                                                                      
2021m12        101.2735             NaN        107.0408        148.5904   
 2022m1         95.0035             NaN        111.0521        129.5658   
 2022m2         98.4584             NaN        116.3363        138.5603   
 2022m3        121.9939             NaN        117.6540        165.9262   
 2022m4         95.2665             NaN        116.9610             NaN   

         MAB_ELE_PRO1100  MAB_ELE_SHP1100  RohiBASEMET1000_org  \
date                                                             
2021m12         123.0767         150.0469             125.2070   
 2022m1         103.1998         120.3381             133.2194   
 2022m2         113.5006         131.5001             138.9056   
 2022m3         133.1330         158.0556             149.8909   
 2022m4         112.9022         134.9355             146.0910   

         RohiENERGY1000_org  RohiMETMIN1000_org  RohiNATGAS1000_org  \
date                                                                  
2021m12            112.3730            116.7152            236.4884   
 2022m1            121.3099            125.2296            196.9111   
 2022m2            131.2732            131.1765            197.5237   
 2022m3            163.1868            141.2833            271.0799   
 2022m4            153.1889            138.0941            243.4360   

         RohCRUDE_PETRO1000_org  RohCOPPER1000_org  WKLWEUR840_org  \
date                                                                 
2021m12                 92.1887           126.7612          1.1304   
 2022m1                106.1731           129.8291          1.1314   
 2022m2                118.3482           131.9636          1.1342   
 2022m3                142.2009           135.7822          1.1019   
 2022m4                130.8354           134.8597          1.0819   

         PRI27840_org  PRI27826_org  PRI27380_org  PRI27250_org  PRI27276_org  \
date                                                                            
2021m12      128.5113           NaN      113.3096      108.1825      115.7489   
 2022m1      131.6285           NaN      115.3906      111.0375      117.8534   
 2022m2      133.3422           NaN      116.4311      112.0571      118.9056   
 2022m3      136.1538           NaN      117.471

In [612]:
df_market.shape

(219, 47)

In [613]:
df_market.info()

<class 'pandas.core.frame.DataFrame'>
Index: 219 entries,  2004m2 to  2022m4
Data columns (total 47 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   MAB_ELE_PRO156          219 non-null    float64
 1   MAB_ELE_SHP156          219 non-null    float64
 2   MAB_ELE_PRO250          219 non-null    float64
 3   MAB_ELE_SHP250          219 non-null    float64
 4   MAB_ELE_PRO276          219 non-null    float64
 5   MAB_ELE_SHP276          219 non-null    float64
 6   MAB_ELE_PRO380          219 non-null    float64
 7   MAB_ELE_SHP380          219 non-null    float64
 8   MAB_ELE_PRO392          219 non-null    float64
 9   MAB_ELE_SHP392          219 non-null    float64
 10  MAB_ELE_PRO756          218 non-null    float64
 11  MAB_ELE_SHP756          218 non-null    float64
 12  MAB_ELE_PRO826          219 non-null    float64
 13  MAB_ELE_SHP826          201 non-null    float64
 14  MAB_ELE_PRO840          219 non-null 

In [614]:
missing_values = df_market.isnull().sum()

features_with_missing_values = missing_values[missing_values > 0]

features_with_missing_values

MAB_ELE_PRO756     1
MAB_ELE_SHP756     1
MAB_ELE_SHP826    18
MAB_ELE_SHP840     1
PRI27826_org      18
PRI27250_org      35
PRI27156_org      23
PRO28756_org       1
PRO271000_org     11
PRO27756_org       1
dtype: int64

In [615]:
df_market.describe().T

count     mean     std     min      25%      50%  \
MAB_ELE_PRO156         219.0000 138.3036 78.8832 16.9407  68.4777 133.5077   
MAB_ELE_SHP156         219.0000 138.3036 78.8832 16.9407  68.4777 133.5077   
MAB_ELE_PRO250         219.0000 104.4319 18.9185 50.7567  93.6135 102.7366   
MAB_ELE_SHP250         219.0000 105.3168 12.7622 64.4207  97.4528 106.0122   
MAB_ELE_PRO276         219.0000 107.4991 11.8619 74.3329 100.5609 108.9923   
MAB_ELE_SHP276         219.0000 114.8984 17.0916 71.7872 103.1498 117.4288   
MAB_ELE_PRO380         219.0000 105.2284 23.5096 34.2134  94.3352 105.0885   
MAB_ELE_SHP380         219.0000 105.7354 19.9482 45.1917  95.9858 107.6958   
MAB_ELE_PRO392         219.0000 111.9481 15.4893 67.5319 103.7400 111.6830   
MAB_ELE_SHP392         219.0000 112.6706 16.8919 64.3723 103.4532 112.5973   
MAB_ELE_PRO756         218.0000  97.8345  8.2415 77.8015  92.4102  97.5731   
MAB_ELE_SHP756         218.0000  94.7849  9.1539 74.6393  88.2210  94.3768   
MAB_ELE_PRO826         219.0000 108.7529 12.0967 61.0480 100.4984 108.9110   
MAB_ELE_SHP826         201.0000  95.9571 12.9462 57.4629  86.1997  97.8689   
MAB_ELE_PRO840         219.0000 109.4183  7.8913 85.9944 104.6000 110.1536   
MAB_ELE_SHP840         218.0000 114.1600 14.6334 79.2402 104.5965 115.7134   
MAB_ELE_PRO1100        219.0000 108.7800 11.8395 69.7866 101.7296 110.3838   
MAB_ELE_SHP1100        219.0000 110.5511 15.7279 71.1589  98.5880 111.9541   
RohiBASEMET1000_org    219.0000  89.7333 20.8101 50.8220  76.5905  88.3904   
RohiENERGY1000_org     219.0000  92.5580 30.6154 31.6323  69.9283  86.2849   
RohiMETMIN1000_org     219.0000  86.0649 21.4108 46.4684  71.9569  85.0702   
RohiNATGAS1000_org     219.0000 103.3678 41.7474 33.9923  72.5948 100.2854   
RohCRUDE_PETRO1000_org 219.0000  89.4465 30.7368 26.6234  66.7040  82.4350   
RohCOPPER1000_org      219.0000  86.3443 23.8115 35.6574  72.3822  88.9860   
WKLWEUR840_org         219.0000   1.2535  0.1202  1.0543   1.1439   1.2400   
PRI27840_org           219.0000 102.1857 10.5241 78.9699  97.3386 103.8836   
PRI27826_org           201.0000 100.1512 10.5494 80.5527  91.8117 102.2518   
PRI27380_org           219.0000 102.1650  4.4549 93.0200  99.0549 103.3209   
PRI27250_org           184.0000 101.9694  2.6469 96.8646 100.2294 102.2687   
PRI27276_org           219.0000 103.1736  6.0764 93.2305  98.7022 103.4374   
PRI27156_org           196.0000  97.6814  3.2359 90.2923  95.0728  97.8530   
PRO28840_org           219.0000 108.6688  9.1324 83.1973 102.8445 109.3037   
PRO281000_org          219.0000 112.4664 12.4112 74.7610 104.8336 113.6166   
PRO28756_org           218.0000  99.7171 10.4680 77.9526  92.9796  98.0512   
PRO28826_org           219.0000 104.3849 13.2227 51.7162  94.5916 105.3743   
PRO28380_org           219.0000 109.4863 22.4910 33.7972  99.7417 112.4906   
PRO28392_org           219.0000 115.0832 17.4083 64.0824 105.4155 115.0922   
PRO28250_org           219.0000 109.1167 21.0501 54.4162  96.5990 106.3959   
PRO28276_org           219.0000 110.4292 14.3517 71.6177 102.0150 111.3414   
PRO27840_org           219.0000 110.9029  8.4446 91.5358 106.6172 109.1566   
PRO271000_org          208.0000 111.1471 12.3212 83.3102 102.7984 110.2844   
PRO27756_org           218.0000  94.5929 13.4650 56.8322  90.2837  97.6202   
PRO27826_org           219.0000 116.6673 13.2053 77.9563 107.4144 116.9943   
PRO27380_org           219.0000  95.4505 29.1954 34.4871  79.1688  88.3780   
PRO27392_org           219.0000 105.5558 12.5925 74.5655  97.5477 104.3663   
PRO27250_org           219.0000  96.8441 16.8578 44.8294  84.7706  95.8238   
PRO27276_org           219.0000 102.1255  9.9599 76.4246  95.6596 103.5496   

                            75%      max  
MAB_ELE_PRO156         198.4739 329.4134  
MAB_ELE_SHP156         198.4739 329.4134  
MAB_ELE_PRO250         114.0909 152.7434  
MAB_ELE_SHP250         115.0305 136.8587  
MAB_ELE_PRO276         115.7358 130.8700  
MAB_ELE_SHP276         

In [616]:
df_sales = pd.read_csv('Data/Case2_Sales data.csv', delimiter=';')
df_sales.head()

DATE Mapped_GCK Sales_EUR
0  01.10.2018         #1         0
1  02.10.2018         #1         0
2  03.10.2018         #1         0
3  04.10.2018         #1         0
4  05.10.2018         #1         0

In [617]:
df_sales.tail()

DATE Mapped_GCK Sales_EUR
9797  23.08.2019        #12         0
9798  23.08.2019        #36      1015
9799  12.08.2019        #12         0
9800  28.08.2019         #8  43763,91
9801  27.08.2019         #8         0

In [618]:
df_sales.shape

(9802, 3)

In [619]:
df_sales.describe()

DATE Mapped_GCK Sales_EUR
count         9802       9802      9802
unique        1216         14      2609
top     16.04.2021         #1         0
freq            14       1179      7134

In [620]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9802 entries, 0 to 9801
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   DATE        9802 non-null   object
 1   Mapped_GCK  9802 non-null   object
 2   Sales_EUR   9802 non-null   object
dtypes: object(3)
memory usage: 229.9+ KB


#### <span style="background-color:#235987; padding:5px; border-radius:5px;">**Sales Data**</span> <a id='sales-data'></a>  
_This section provides a compreensive understanding of the Sales data and its preprocessing steps._

##### Click [here](#table-of-contents) ⬆️ to return to the Index.

> Functions

In [621]:
#plot the sales data over a period of time
def plot_sales(dataframe, x_column, y_column, title, x_label, y_label):
    
    fig = px.line(dataframe, 
                  x=x_column, 
                  y=y_column,
                  title=title, 
                  labels={x_column: x_label, y_column: y_label})

    fig.show()

In [622]:
#plot the sales data by product
def plot_sales_by_product(dataframe, x_column, y_column, title, x_label, y_label):
    
    fig = px.bar(dataframe, 
                 x=x_column, 
                 y=y_column, 
                 title=title, 
                 labels={x_column: x_label, y_column: y_label},
                 text=y_column)
    
    fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
    fig.update_layout(xaxis=dict(type='category', tickmode='linear'))
    fig.show()

In [623]:
#plot the sales data over a period of time by product
def plot_sales_by_product_month(dataframe, x_column, y_column, category_column, title, x_label, y_label):
    
    fig = px.line(dataframe, 
                  x=x_column, 
                  y=y_column, 
                  color=category_column,  # Different lines for each product
                  title=title, 
                  labels={x_column: x_label, y_column: y_label, category_column: "Product"})
    
    fig.update_layout(xaxis=dict(tickangle=-45))  # Rotate x-axis labels for better readability
    fig.show()

In [624]:
def plot_sales_by_product_month_comparison(positive_sales, negative_sales, x_column, y_column, category_column, title, x_label, y_label):
    
    for product in positive_sales[category_column].unique():
        # Filter data for the current product (both positive and negative sales)
        df_pos = positive_sales[positive_sales[category_column] == product]
        df_neg = negative_sales[negative_sales[category_column] == product]
        
        # Create a figure for the current product
        fig = go.Figure()

        # Plot positive sales for the current product
        fig.add_trace(go.Scatter(x=df_pos[x_column], 
                                 y=df_pos[y_column], 
                                 mode='lines', 
                                 name=f'Positive Sales - {product}', 
                                 line=dict(color='green')))

        # Plot negative sales (absolute value) for the current product
        fig.add_trace(go.Scatter(x=df_neg[x_column], 
                                 y=df_neg[y_column],  # Negative sales as positive
                                 mode='lines', 
                                 name=f'Negative Sales (abs) - {product}', 
                                 line=dict(color='red')))

        # Update the layout of the plot
        fig.update_layout(
            title=f'{title} - {product}', 
            xaxis_title=x_label, 
            yaxis_title=y_label,
            showlegend=True,
            xaxis=dict(tickangle=-45)  # Rotate x-axis labels for readability
        )

        # Show the plot
        fig.show()

> Dtypes

In [625]:
# Fixing dtypes of df_sales
df_sales['DATE'] = pd.to_datetime(df_sales['DATE'], dayfirst=True)
df_sales['Sales_EUR'] = df_sales['Sales_EUR'].str.replace(',', '.').astype(float)

# DroppinG the # in values of the column 'Mapped_GCK'
df_sales['Mapped_GCK'] = df_sales['Mapped_GCK'].str.replace('#', '')
df_sales['Mapped_GCK'] = df_sales['Mapped_GCK'].astype(int)

In [626]:
# Mapped_GCK categories by frequecy
df_sales['Mapped_GCK'].value_counts()

Mapped_GCK
1     1179
3     1017
5      959
8      944
4      877
12     803
6      794
16     765
11     732
13     441
36     434
9      333
20     293
14     231
Name: count, dtype: int64

In [627]:
# Drop rows in df_sales with Sales_EUR = 0
df_sales = df_sales[df_sales['Sales_EUR'] != 0]

In [628]:
# Rows with values for Sales
df_sales.shape[0]

2668

In [629]:
# Max and min values of Sales
df_sales['Sales_EUR'].max(), df_sales['Sales_EUR'].min()

(41127988.02, -506381.17)

>Negative Sales' values Analysis

In [630]:
negative_sales = df_sales[df_sales['Sales_EUR'] < 0]
negative_sales 

DATE  Mapped_GCK  Sales_EUR
90   2022-02-02           6   -54.0000
96   2022-04-04          11  -587.5900
102  2021-12-02           3  -183.5000
130  2022-03-02           6   -93.8800
155  2021-05-03           5  -166.2400
...         ...         ...        ...
9333 2019-09-03          20    -6.6000
9466 2018-11-05           4 -2029.9100
9487 2019-04-30          13   -26.3100
9681 2020-07-31           9   -35.2500
9775 2019-03-26           8 -1000.0000

[276 rows x 3 columns]

<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

**Negative values for Sales** 
There seems to be no evident pattern for the negative sales values. These might be:
- **Dataset errors**: errors that occured during the recording of sales records.
- **Production errors**: errors that occured in the production of a certain product or product group.
- **Returns**: customer returns are associated with a refund, creating a negative cash flow for the company to offset the positive cash flow generated at the time of purchase.


In [631]:
negative_sales['Year_Month'] = negative_sales['DATE'].dt.to_period('M')
grouped_negative_sales_month = negative_sales.groupby('Year_Month')['Sales_EUR'].sum().reset_index()

grouped_negative_sales_month['Year_Month'] = grouped_negative_sales_month['Year_Month'].astype(str)

plot_sales(grouped_negative_sales_month, 'Year_Month', 'Sales_EUR', 'Negative Sales by Month', 'Year-Month', 'Total Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\267356584.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

At first sight, no evident pattern emerged from the **Negative Sales by Month** plot. However, the analysis highlights January 2020 as a significant outlier, displaying a much lower value than other months, which may indicate production errors in a specific product group. Excluding this outlier, a subtle pattern can be observed: sales tend to drop noticeably twice a year—once in the first half and once in the second—though the exact months vary.

In [632]:
negative_sales['Day'] = negative_sales['DATE'].dt.day
grouped_negative_sales_day = negative_sales.groupby('Day')['Sales_EUR'].sum().reset_index()

grouped_negative_sales_day['Day'] = grouped_negative_sales_day['Day'].astype(str)

plot_sales(grouped_negative_sales_day, 'Day', 'Sales_EUR', 'Negative Sales by Day', 'Day', 'Total Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\812041059.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

Considering the **Negative Sales by Day** plot, there seems to be a tendency of much lower values for negative sales in the first days of the month, which slows down until the last days. Besides the first days, around the 20th day of the month there is a new, but not as obvious, peak.

In [633]:
grouped_negative_sales_products= negative_sales.groupby('Mapped_GCK')['Sales_EUR'].sum().reset_index()

plot_sales_by_product(grouped_negative_sales_products, 'Mapped_GCK', 'Sales_EUR', 'Negative Sales by Product', 'Product Group', 'Total Negative Sales (EUR)')

<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

Product Groups with **most negative sales**, from the lowest value: Group **1**, Group **5**, Group **3**, Group **11**, Group **6**, Group **16**.
- Are this products with the most sales as well? Is the negative sales value proportional to the positive values?

In [634]:
negative_sales['Year_Month'] = negative_sales['DATE'].dt.to_period('M').astype(str)
grouped_negative_sales_prod_month = negative_sales.groupby(['Year_Month', 'Mapped_GCK'])['Sales_EUR'].sum().reset_index()

plot_sales_by_product_month(grouped_negative_sales_prod_month, 'Year_Month', 'Sales_EUR', 'Mapped_GCK', 
                            'Negative Sales by Product and Month', 'Year-Month', 'Total Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\499746840.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

The visualization of **Negative Sales by Poduct, by month** gives us an idea of the main drivers for the lower peaks in sales values. 
- **Mar 2019**: most negative sales for this month were driven by the product group **11**, making it possibly a production error. 
- **Jan 2020**: the month with the lowest value for sales resulted from a variety of product groups having a peak in negative sales, namely group **6**, group **16**, group **3**, group **5** and group **1**.
- **Nov 2020**: while this was a month that looked within normal limits, now its possible to identify that the vast majority of negative sales for November 2020 come from the product group **3**, making it very likely a production error. 
- **Apr 20201**: most negative sales for this month were driven by the product group **5**, with a slight peak in group **1** as well, making it possibly a production error.
- **Sep and Oct 2021**: both this months have a similar behaviour to the one of **Apr 2021**, being the main drivers of negative sales products **5** and **1**.

>Positive Sales' values Analysis

In [635]:
positive_sales = df_sales[df_sales['Sales_EUR'] > 0]
positive_sales

DATE  Mapped_GCK     Sales_EUR
26   2018-10-24           1  1702929.2100
31   2021-08-20          16   111107.3800
32   2021-08-17          16     3685.0000
43   2018-10-29           1   974208.5500
45   2018-10-31           1 33421781.0300
...         ...         ...           ...
9788 2019-08-23           3  2695566.6400
9795 2019-08-23          13     3000.0000
9796 2019-08-28          14     1602.0000
9798 2019-08-23          36     1015.0000
9800 2019-08-28           8    43763.9100

[2392 rows x 3 columns]

In [636]:
positive_sales['Year_Month'] = positive_sales['DATE'].dt.to_period('M')
grouped_positive_sales_month = positive_sales.groupby('Year_Month')['Sales_EUR'].sum().reset_index()

grouped_positive_sales_month['Year_Month'] = grouped_positive_sales_month['Year_Month'].astype(str)

plot_sales(grouped_positive_sales_month, 'Year_Month', 'Sales_EUR', 'Positive Sales by Month', 'Year-Month', 'Total Positive Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\2664286345.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [637]:
# comparing the percentage change in sales for positive and negative sales
grouped_positive_sales_month['Sales_EUR_pct_change'] = round(grouped_positive_sales_month['Sales_EUR'].pct_change() * 100,2)
grouped_negative_sales_month['Sales_EUR_pct_change'] = round(grouped_negative_sales_month['Sales_EUR'].pct_change() * 100,2)

merged_sales = pd.merge(grouped_positive_sales_month[['Year_Month', 'Sales_EUR_pct_change']], 
                        grouped_negative_sales_month[['Year_Month', 'Sales_EUR_pct_change']], 
                        on='Year_Month', 
                        suffixes=('_positive', '_negative'))

merged_sales['Year_Month'] = merged_sales['Year_Month'].astype(str)

table = merged_sales.set_index('Year_Month').T

table


Year_Month                     2018-11   2018-12  2019-01  2019-02  2019-03  \
Sales_EUR_pct_change_positive -37.8000  100.1300 -16.3500   5.6400  13.9000   
Sales_EUR_pct_change_negative      NaN 2510.1400 -44.0600 -46.8900 578.0800   

Year_Month                     2019-04  2019-05  2019-06  2019-07  2019-08  \
Sales_EUR_pct_change_positive  -1.7900   7.8800   1.1300  -2.0100  -4.9700   
Sales_EUR_pct_change_negative -50.4700 -62.4300  51.0100  -5.3200  43.1900   

Year_Month                     2019-09  2019-10  2019-11  2019-12  2020-01  \
Sales_EUR_pct_change_positive  28.4100 -19.4900  -2.6200   6.3700 -16.0300   
Sales_EUR_pct_change_negative  52.9900  52.7200 -83.3100 554.8600  60.3400   

Year_Month                     2020-02  2020-03  2020-04  2020-05  2020-06  \
Sales_EUR_pct_change_positive   6.2100  11.4400  -7.9500  -9.1800  17.2400   
Sales_EUR_pct_change_negative -85.3100 119.7000 -62.5000 -88.8800 669.0000   

Year_Month                     2020-07  2020-08  2020-09    2020-10   2020-11  \
Sales_EUR_pct_change_positive -10.2600  10.8700  17.7900   -23.0600   13.6900   
Sales_EUR_pct_change_negative -51.3600 -50.4600 -98.6700 13242.7000 1281.3800   

Year_Month                     2020-12  2021-01  2021-02  2021-03  2021-04  \
Sales_EUR_pct_change_positive -14.0200 -41.8500 120.1800 -13.3100   4.9200   
Sales_EUR_pct_change_negative -93.1300 -91.7300 266.8500 359.6200 719.9000   

Year_Month                     2021-05  2021-06  2021-07  2021-08  2021-09  \
Sales_EUR_pct_change_positive  -7.1700   5.9400  -2.1100  15.0700   3.5400   
Sales_EUR_pct_change_negative -84.6100 -83.3700 154.8500  73.2700 736.5600   

Year_Month                     2021-10  2021-11  2021-12  2022-01  2022-02  \
Sales_EUR_pct_change_positive  -5.7600  -6.4600  -8.1100   1.8900   0.5300   
Sales_EUR_pct_change_negative -11.9500 -63.8800  18.1800  80.1800 -56.8700   

Year_Month                     2022-03   2022-04  
Sales_EUR_pct_change_positive  18.5100  -12.1900  
Sales_EUR_pct_change_negative -93.1300 2289.1600

<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

**Nevative vs Positive Sales Values**

- (*Outlier*) **November 2018** was a month of few activity, with the lowest overall positive sales and negative sales close to non.

- **September 2019** is a peak in positive sales, that is followed by a peak in negative sales, in **October 2019** - which might come has a consequence of the high sales in the previous month. 

- **January 2020** is not only the month with highest negative values, but it is also a month categorized by the bellow average positive sales, which is likely evidence of a outside event's affect in the sales behavior.

- (*Outlier*) **January 2021** is a month marked by a abrupt change in sales behavior, it is also a month of few activity, with one of the lowest sales value, by far.

- **September 2021** is a month that peaks in both negative and positive sales values - this kind of situation can be taken as a proportional increase in positive and negative (absolute) sales.

In [638]:
positive_sales['Day'] = positive_sales['DATE'].dt.day
grouped_positive_sales_day = positive_sales.groupby('Day')['Sales_EUR'].sum().reset_index()

grouped_positive_sales_day['Day'] = grouped_positive_sales_day['Day'].astype(str)

plot_sales(grouped_positive_sales_day, 'Day', 'Sales_EUR', 'Positive Sales by Day', 'Day', 'Total Positive Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\3740950382.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

Similar to Negative sales values, the positive sales peaks at the beginning of the month. This aligns with the expected spending behavior coming from enterprises, due to budgeting/cash flow management, payroll and other expenses settlement at the end of the month, between others. 

In [639]:
grouped_positive_sales_product= positive_sales.groupby('Mapped_GCK')['Sales_EUR'].sum().reset_index()

plot_sales_by_product(grouped_positive_sales_product, 'Mapped_GCK', 'Sales_EUR', 'Positive Sales by Product', 'Product Group', 'Total Positive Sales (EUR)')

<span style="background-color:#235987; padding:5px; border-radius:5px;">**Remarks**</span> <a id='sales-data'></a>

Products positive sales values, by order:

**1.** Product 1

**2.** Product 3

**3.** Product 5

**4.** Product 11

**5.** Product 8

**6.** Product 6

**7.** Product 16

**8.** Product 4

**9.** Product 12

**10.** Product 36

**11.** Product 13

**12.** Product 14

**13.** Product 9

**14.** Product 20

In [640]:
positive_sales['Year_Month'] = positive_sales['DATE'].dt.to_period('M').astype(str)
grouped_positive_sales_prod_month = positive_sales.groupby(['Year_Month', 'Mapped_GCK'])['Sales_EUR'].sum().reset_index()

plot_sales_by_product_month(grouped_positive_sales_prod_month, 'Year_Month', 'Sales_EUR', 'Mapped_GCK', 
                            'Positive Sales by Product and Month', 'Year-Month', 'Total Positive Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\1474112446.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [641]:
negative_sales['Sales_EUR']= negative_sales['Sales_EUR'].abs()
negative_sales['Year_Month'] = negative_sales['DATE'].dt.to_period('M').astype(str)
grouped_neg_abs_sales_prod_month = negative_sales.groupby(['Year_Month', 'Mapped_GCK'])['Sales_EUR'].sum().reset_index()

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\1125194768.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\1125194768.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



> Total Sales

In [642]:
df_sales['Year_Month'] = df_sales['DATE'].dt.to_period('M').astype(str)

df_sales = df_sales.groupby(['Year_Month', 'Mapped_GCK'])['Sales_EUR'].sum().reset_index()

In [643]:
# Creating new columns for each product
df_p1 = df_sales[df_sales['Mapped_GCK'] == 1]
df_p3 = df_sales[df_sales['Mapped_GCK'] == 3]
df_p4 = df_sales[df_sales['Mapped_GCK'] == 4]
df_p5 = df_sales[df_sales['Mapped_GCK'] == 5]
df_p6 = df_sales[df_sales['Mapped_GCK'] == 6]
df_p8 = df_sales[df_sales['Mapped_GCK'] == 8]
df_p9 = df_sales[df_sales['Mapped_GCK'] == 9]
df_p11 = df_sales[df_sales['Mapped_GCK'] == 11]
df_p12 = df_sales[df_sales['Mapped_GCK'] == 12]
df_p13 = df_sales[df_sales['Mapped_GCK'] == 13]
df_p14 = df_sales[df_sales['Mapped_GCK'] == 14]
df_p16 = df_sales[df_sales['Mapped_GCK'] == 16]
df_p20 = df_sales[df_sales['Mapped_GCK'] == 20]
df_p36 = df_sales[df_sales['Mapped_GCK'] == 36]

df_p1 = df_p1.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p3 = df_p3.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p4 = df_p4.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p5 = df_p5.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p6 = df_p6.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p8 = df_p8.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p9 = df_p9.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p11 = df_p11.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p12 = df_p12.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p13 = df_p13.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p14 = df_p14.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p16 = df_p16.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p20 = df_p20.groupby('Year_Month')['Sales_EUR'].sum().reset_index()
df_p36 = df_p36.groupby('Year_Month')['Sales_EUR'].sum().reset_index()

df_sales = df_sales.merge(df_p1[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 1'}), on='Year_Month')
df_sales = df_sales.merge(df_p3[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 3'}), on='Year_Month')
df_sales = df_sales.merge(df_p4[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 4'}), on='Year_Month')
df_sales = df_sales.merge(df_p5[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 5'}), on='Year_Month')
df_sales = df_sales.merge(df_p6[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 6'}), on='Year_Month')
df_sales = df_sales.merge(df_p8[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 8'}), on='Year_Month')
df_sales = df_sales.merge(df_p9[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 9'}), on='Year_Month')
df_sales = df_sales.merge(df_p11[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 11'}), on='Year_Month')
df_sales = df_sales.merge(df_p12[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 12'}), on='Year_Month')
df_sales = df_sales.merge(df_p13[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 13'}), on='Year_Month')
df_sales = df_sales.merge(df_p14[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 14'}), on='Year_Month')
df_sales = df_sales.merge(df_p16[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 16'}), on='Year_Month')
df_sales = df_sales.merge(df_p20[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 20'}), on='Year_Month')
df_sales = df_sales.merge(df_p36[['Year_Month', 'Sales_EUR']].rename(columns={'Sales_EUR': 'Product 36'}), on='Year_Month')

In [644]:
df_sales['Total_Sales_EUR'] = df_sales.groupby('Year_Month')['Sales_EUR'].transform('sum')

df_sales.drop(columns=['Sales_EUR', 'Mapped_GCK'], inplace=True)

In [645]:
df_sales.drop_duplicates(inplace=True)
df_sales.reset_index(drop=True, inplace=True)

In [646]:
df_sales

Year_Month     Product 1     Product 3   Product 4     Product 5  \
0     2018-10 36098918.7900  8089465.9600 397760.6900  2499061.1900   
1     2018-11  5140760.0000 11863001.5100 371322.4200  8993944.0400   
2     2019-01 27728148.3500 10705300.6300 484173.8800  8233205.0700   
3     2019-02 34793163.5300 10167796.8600 620031.8000  6879250.9900   
4     2019-03 33856803.0400 11040544.5700 491643.5900 11914313.6300   
5     2019-05 38897974.2900 12798945.1800 375671.5700 11389673.2600   
6     2019-06 39106750.2300 11546452.1900 190721.1500 13681879.1600   
7     2019-07 37912572.3000 13421550.8300 381274.9600 11145039.0300   
8     2019-08 38042677.0500 11359418.3000 301973.4300  7645745.2900   
9     2019-09 40602780.6400 16579668.1100 443349.3200 15018604.6300   
10    2019-10 39012299.1900 11382774.4500  93226.3200 11542647.4700   
11    2019-11 37323903.0700 13826060.4800 298011.6800  8104241.6900   
12    2020-01 34116651.4700 11699896.3400 218623.4800  7184359.3100   
13    2020-03 39064430.8800 12281515.4600 197670.4800  9232237.6800   
14    2020-04 37981531.3200 12008066.8500 387939.9800  7990825.3200   
15    2020-05 31400212.5400 10708632.1800 315488.9800  9872426.5700   
16    2020-06 30896804.0700 13321842.3600 329430.9600 15873894.9100   
17    2020-07 35212730.1300 12774946.2900 294293.7900  5590709.7100   
18    2020-08 37482085.5800 11373503.0800 172952.2400 11463705.6400   
19    2020-09 38879360.2000 15912683.1700 645058.0000 14160091.0500   
20    2020-10 41453854.5000 10003848.4400 332603.8400  4647204.3100   
21    2020-11 37643611.5700 15021617.7900 413558.1400 10164873.5700   
22    2020-12 34816959.1800  9755535.9900 303249.5900  8845742.3000   
23    2021-02 34630522.8100 16136114.3800 807758.7900 16650952.7200   
24    2021-03 33476138.6400 14128552.8000 451955.4100 11102204.5800   
25    2021-04 41257397.9000 13461271.9300 295958.9200  7085482.5200   
26    2021-05 37760015.2800 10298079.4100 573707.2700  8767595.7300   
27    2021-06 35801398.4700 13650393.4500 829442.3300 11687229.0300   
28    2021-07 37130004.2100 13288195.2400 434669.6100  8321886.3800   
29    2021-10 44483013.8600 13797043.9800 344985.7300 11472426.0100   
30    2021-11 34187395.6000 18686819.8500 290506.8700  9606371.1400   
31    2021-12 33950069.8300 12376414.5800 199188.7500  8656479.6400   
32    2022-02 36293076.3300 15665217.4700 245306.1300  7675455.5500   
33    2022-03 40220407.0300 14320304.5500 406399.8300 14739187.1800   
34    2022-04 41053203.1200 11999374.5100 302057.0400  8884607.5300   

     Product 6    Product 8  Product 9   Product 11  Product 12  Product 13  \
0  369231.6000  586052.7400  3219.3200 1021303.5000  28686.3300  27666.1000   
1  473046.9600  526292.7700  1875.9000 1898844.8000   1070.0000  68180.0000   
2  598874.1000  381400.1500  1487.0000 2216391.7400  21484.0000  27198.2900   
3  542037.5200  368475.5700  3234.2800  610456.6000  34214.7400  32638.6300   
4  735604.2300  591288.3600 13050.6100  493999.8100  10118.0000  11890.0000   
5  315978.7700  298806.8400  4614.0800 1066063.9000  24468.0000   2550.3100   
6  282963.1900  420707.7300  4488.0900  492133.2000  44255.6800  34849.2100   
7  725411.9900  366980.2600  1766.4000  585648.1400  21318.4800  43008.3500   
8  206931.1700  397537.2000  6371.5000 2898779.8700  19217.0000  22497.6000   
9  499147.7300 1084685.9200 13930.2600 3749980.4600 292490.4000  19601.0500   
10 581767.8500  111791.3400   555.0000  218976.7800 140819.6900  65754.5500   
11 181159.2000  542923.4200 24181.8000 1194770.7500 320942.2800  43190.7600   
12 -15202.9600  323769.2000  7647.5000  584370.0000 213315.3200  19519.5900   
13 529808.9400  845857.7600  2233.0000 2969948.6500 185833.8900  23513.5900   
14 782042.9800  302765.6800  1476.8500  813303.0100  79632.0000  30716.1100   
15 750367.7000  712746.3800  3764.1200  842526.1000 153949.2400   3762.8200   
16 526425.1800  599831.9300  7555.4500 2246948.0000 172712.8800   9612.8000   
17 730187.6800  583157.6

#### <span style="background-color:#000027; padding:5px; border-radius:5px;">**1. Data Loading and Description**</span> <a id='data-loading-and-description'></a>  
_This section provides an overview of the datasets, including their structure, size, and general characteristics._  

Click [here](#table-of-contents) ⬆️ to return to the Index.


##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**1.1. Sales Dataset**</span> <a id='sales-dataset'></a>

In [647]:
df_sales = pd.read_csv('Data/Case2_Sales data.csv', delimiter=';')
df_sales.head()

DATE Mapped_GCK Sales_EUR
0  01.10.2018         #1         0
1  02.10.2018         #1         0
2  03.10.2018         #1         0
3  04.10.2018         #1         0
4  05.10.2018         #1         0

First thing: we got the daily sales, so we will need to group this into monthly sales per product.

In [648]:
df_sales.shape

(9802, 3)

##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**1.2. Consumer Index Prices Dataset**</span> <a id='cpi-dataset'></a>

In [649]:
df_cpi = pd.read_excel('Data/CPI.xlsx')
df_cpi.head()

c:\Users\marga\Documents\anaconda\Lib\site-packages\openpyxl\styles\stylesheet.py:226: UserWarning:

Workbook contains no default style, apply openpyxl's default



Consumer price index: Germany, months Unnamed: 1            Unnamed: 2  \
0      Consumer price index for Germany        NaN                   NaN   
1                               Germany        NaN                   NaN   
2                          Year\nMonths        NaN  Consumer price index   
3                                   NaN        NaN              2020=100   
4                                  2018    January               96.4000   

  Unnamed: 3                       Unnamed: 4 Unnamed: 5  \
0        NaN                              NaN        NaN   
1        NaN                              NaN        NaN   
2        NaN  Change on previous year's month        NaN   
3        NaN                           in (%)        NaN   
4          e                           1.4000          e   

                 Unnamed: 6 Unnamed: 7  
0                       NaN        NaN  
1                       NaN        NaN  
2  Change on previous month        NaN  
3                    in (%)        NaN  
4                   -0.5000          e

##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**1.3. Market Dataset**</span> <a id='market-dataset'></a>

In [650]:
df_market = pd.read_excel('Data/Case2_Market data.xlsx', sheet_name="Original Values", index_col=0, header=2)
df_market.head()

MAB_ELE_PRO156  MAB_ELE_SHP156  MAB_ELE_PRO250  MAB_ELE_SHP250  \
date                                                                      
 2004m2         16.9407         16.9407        112.0913         83.4589   
 2004m3         23.7119         23.7119        136.3280        106.1682   
 2004m4         24.4352         24.4352        117.7918         92.0076   
 2004m5         23.7081         23.7081        109.0025         85.6965   
 2004m6         27.0091         27.0091        133.7857        106.6415   

         MAB_ELE_PRO276  MAB_ELE_SHP276  MAB_ELE_PRO380  MAB_ELE_SHP380  \
date                                                                      
 2004m2         82.6230         79.4525        124.2896         86.5605   
 2004m3        100.5566         97.0129        143.4117        106.3445   
 2004m4         89.6532         84.9324        129.0838         95.5797   
 2004m5         86.8806         82.3728        135.5904        100.0870   
 2004m6         99.0108         95.1087        136.4249        110.8897   

         MAB_ELE_PRO392  MAB_ELE_SHP392  MAB_ELE_PRO756  MAB_ELE_SHP756  \
date                                                                      
 2004m2        109.3340        110.4953         91.2219         89.9873   
 2004m3        140.8846        144.6862         85.8663         79.8836   
 2004m4        105.8536        102.6558         85.6225         79.7408   
 2004m5        101.8648        100.3053         85.3787         79.5980   
 2004m6        120.3329        119.6164         85.1349         79.4552   

         MAB_ELE_PRO826  MAB_ELE_SHP826  MAB_ELE_PRO840  MAB_ELE_SHP840  \
date                                                                      
 2004m2        111.3538         73.6013        107.6014         79.2402   
 2004m3        127.5586         84.0476        110.1874         98.6190   
 2004m4        108.7323         73.0260        108.1666         89.7740   
 2004m5        110.6452         74.5919        108.4259         87.4638   
 2004m6        122.0210         82.3433        110.5699         97.3645   

         MAB_ELE_PRO1100  MAB_ELE_SHP1100  RohiBASEMET1000_org  \
date                                                             
 2004m2          97.1229          80.0985              54.0398   
 2004m3         113.7839          96.0159              54.6662   
 2004m4         101.7152          85.1672              54.8727   
 2004m5         101.2757          84.4858              51.2304   
 2004m6         112.0572          96.9633              52.8763   

         RohiENERGY1000_org  RohiMETMIN1000_org  RohiNATGAS1000_org  \
date                                                                  
 2004m2             44.1233             48.7479             87.0770   
 2004m3             47.5890             49.2562             87.1927   
 2004m4             47.7790             49.4238             91.3799   
 2004m5             53.5909             46.4684             99.0445   
 2004m6             50.7996             47.8039             98.6363   

         RohCRUDE_PETRO1000_org  RohCOPPER1000_org  WKLWEUR840_org  \
date                                                                 
 2004m2                 39.6395            36.6238          1.2646   
 2004m3                 42.5920            39.9311          1.2262   
 2004m4                 42.6506            39.1349          1.1985   
 2004m5                 47.5171            36.2784          1.2007   
 2004m6                 44.9676            35.6574          1.2138   

         PRI27840_org  PRI27826_org  PRI27380_org  PRI27250_org  PRI27276_org  \
date                                                                            
 2004m2       78.9699       80.7574       93.0200           NaN       93.2305   
 2004m3       79.6736       80.9621       93.5403           NaN       93.3357   
 2004m4       80.3376       80.7574       93.8524           NaN       93.4409   
 2004m5       80.7988       80.7574       93.852

#### <span style="background-color:#000027; padding:5px; border-radius:5px;">**2. Sales Data Analysis**</span> <a id='sales'></a>  
_This section provides an overview of the datasets, including their structure, size, and general characteristics._  

Click [here](#table-of-contents) ⬆️ to return to the Index.

---

##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**2.1. Data Types**</span> <a id='Data-Types'></a>  
_By reducing memory consumption, we can enhance the performance of subsequent sections, including **feature engineering and modeling algorithms**._   

Click [here](#table-of-contents) ⬆️ to return to the Index.

In [651]:
df_sales.dtypes

DATE          object
Mapped_GCK    object
Sales_EUR     object
dtype: object

**Note:** The columns datatyes are not optimize, in data preparation is important to convert:
- `Date` to dd-mm-yyyy (datetime format)
- `Sales_EUR` to float
- `Mapped_GCK` to int (removing the character #)

Right now we have a memory usage of 229.9+ KB, but we can clearly see that these three columns can be optimized.

In order to facilitate the understanding of the data, **dtypes are going to be fixed** at this stage.

In [652]:
# convert DATE to datetime format
df_sales['DATE'] = pd.to_datetime(df_sales['DATE'], dayfirst=True)

In [653]:
# create the feature 'month_year', that stores the month and year of the corresponding date value
df_sales['month_year'] = pd.to_datetime(df_sales['DATE'], format='%d.%m.%Y').dt.to_period('M')

In [654]:
# convert Sales_EUR to float
df_sales['Sales_EUR'] = df_sales['Sales_EUR'].str.replace(',', '.').astype(float)

In [655]:
# convert Mapped_GCK
df_sales['Mapped_GCK'] = df_sales['Mapped_GCK'].str.replace("#", "").astype(int)

In [656]:
df_sales.dtypes

DATE          datetime64[ns]
Mapped_GCK             int32
Sales_EUR            float64
month_year         period[M]
dtype: object

`Note`: The optimization function, which optimizes the specific bytes of the data type, will be applied in the preparation pipeline.

---

##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**2.2. Descriptive Summary**</span> <a id='Descriptive-Summary'></a>  
_A detailed summary of the variables, including their central tendency, dispersion, and distribution._  

Click [here](#table-of-contents) ⬆️ to return to the Index.

In [657]:
# Summary statistics for numeric columns
pd.set_option('display.float_format', lambda x: f'{x:.4f}')
df_sales.describe()

DATE  Mapped_GCK     Sales_EUR
count                           9802   9802.0000     9802.0000
mean   2020-07-19 17:55:48.785961984      9.0007   272785.6867
min              2018-10-01 00:00:00      1.0000  -506381.1700
25%              2019-08-30 00:00:00      4.0000        0.0000
50%              2020-07-26 12:00:00      8.0000        0.0000
75%              2021-06-09 00:00:00     12.0000        0.0000
max              2022-04-30 00:00:00     36.0000 41127988.0200
std                              NaN      7.6243  2230789.1619

In [658]:
# Check unique products
print('Unique products:', list(df_sales['Mapped_GCK'].unique()))

Unique products: [1, 11, 6, 8, 12, 16, 4, 5, 3, 9, 14, 13, 20, 36]


- We have a total of 14 different products, and our approach will involve individual treatment for each, including different feature selection and modeling.

- There is a negative minimum value for sales, which may indicate refunds or adjustments. We will investigate this further.

- Considering the minimum and maximum values of the sales column, we observe a precision of two decimal places and a maximum value of 40M. This suggests that the float64 data type can be optimized.

- The columns in this dataframe do not contain any missing values.

---
##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**2.3. Inconsistencies**</span> <a id='inconsistenciesDU'></a>  
_Checking for inconsistencies in data, such as incorrect formats, out-of-range values, or logical errors._ 

Click [here](#table-of-contents) ⬆️ to return to the Index.

- `Sales_EUR` == 0:

In [659]:
len(df_sales['Sales_EUR'] == 0)

9802

We have 9,802 rows where sales are 0, indicating 'no sales' on those days. Since we will be grouping by month, this won't have a significant impact. However, for now, we will drop these rows to make the dataframe more manageable.

In [660]:
df_sales.drop(df_sales[df_sales['Sales_EUR'] == 0].index, inplace=True)

- `Sales_EUR` Negative Values: Sales_EUR has entries that are negative values. How do they behave? Are they impactful?

In [661]:
# Min and max values of Sales_EUR
df_sales['Sales_EUR'].min(), df_sales['Sales_EUR'].max()

(-506381.17, 41127988.02)

##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**2.4. Negative Values Analysis**</span> <a id='negative-values'></a> 

>**Functions**

In [662]:
#plot the sales data over a period of time
def plot_sales(dataframe, x_column, y_column, title, x_label, y_label):
    fig = px.line(dataframe, x=x_column, y=y_column,title=title, 
                  labels={x_column: x_label, y_column: y_label})

    fig.show()

In [663]:
#plot the sales data by product
def plot_sales_by_product(dataframe, x_column, y_column, title, x_label, y_label):
    fig = px.bar(dataframe, x=x_column, y=y_column, title=title, 
                labels={x_column: x_label, y_column: y_label},
                text=y_column)
    
    fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
    fig.update_layout(xaxis=dict(type='category', tickmode='linear'))
    fig.show()

In [664]:
#plot the sales data over a period of time by product
def plot_sales_by_product_month(dataframe, x_column, y_column, category_column, title, x_label, y_label):
    
    fig = px.line(dataframe, 
                  x=x_column, 
                  y=y_column, 
                  color=category_column,  # Different lines for each product
                  title=title, 
                  labels={x_column: x_label, y_column: y_label, category_column: "Product"})
    
    fig.update_layout(xaxis=dict(tickangle=-45))  # Rotate x-axis labels for better readability
    fig.show()

In [665]:
def plot_sales_trends(data, trend_type='linear'):

    data['month_year'] = pd.to_datetime(data['month_year'])
    unique_products = data['Mapped_GCK'].unique()
    
    fig = make_subplots(
        rows=1, cols=len(unique_products),
        subplot_titles=[f'Product {p}' for p in unique_products],
        horizontal_spacing=0.02
    )
    
    for i, product in enumerate(unique_products):
        product_data = data[data['Mapped_GCK'] == product].copy()
        product_data = product_data.sort_values(by='month_year')
        
        if trend_type == 'linear':
            product_data['month_numeric'] = product_data['month_year'].map(pd.Timestamp.toordinal)
            slope, intercept, _, _, _ = linregress(product_data['month_numeric'], product_data['Sales_EUR'])
            product_data['Trend'] = intercept + slope * product_data['month_numeric']
        else:  # Rolling mean trend
            product_data['Trend'] = product_data['Sales_EUR'].rolling(window=3, min_periods=1).mean()
        
        fig.add_trace(
            go.Scatter(x=product_data['month_year'], y=product_data['Sales_EUR'],
                       mode='lines+markers', name=f'Product {product} Sales',
                       line=dict(color='blue')),
            row=1, col=i+1)
        
        fig.add_trace(
            go.Scatter(x=product_data['month_year'], y=product_data['Trend'],
                       mode='lines', name=f'Product {product} Trend',
                       line=dict(color='red', dash='dot')),
            row=1, col=i+1)
    
    fig.update_layout(
        title="Positive Sales Trends by Product (Monthly)",
        xaxis_title="Year-Month",
        yaxis_title="Total Positive Sales (EUR)",
        showlegend=False,
        height=500, width=500 * len(unique_products),
        xaxis=dict(tickangle=-45)
    )
    
    fig.show()

> Vizualizations

In [666]:
negative_sales = df_sales[df_sales['Sales_EUR'] < 0]

`Remarks`  

**Negative Values for Sales**  
There seems to be no evident pattern for the negative sales values. These might be due to:  

- **Dataset Errors**: Errors that occurred during the recording of sales records, such as mistakenly imputing a "-", could be a possibility. However, since we have evidence that the system is automated — for instance, all days are recorded even when there are no sales — we can likely exclude human errors.  

- **Production Errors Leading to Returns**: Products in the infrastructure sector often involve large, complex systems. Customers might return products due to:

    - Defective Components discovered upon installation.
    - Wrong Specifications delivered for projects.

Returns are typically associated with Refunds or Credit Notes issued after validating customer complaints or issues, resulting in a negative cash flow for the company to offset the positive cash flow generated at the time of purchase.  

- **Inventory Reconciliations** where incorrect sales entries are adjusted.

---


In [667]:
grouped_negative_sales_month = negative_sales.groupby('month_year')['Sales_EUR'].sum().reset_index()

grouped_negative_sales_month['month_year'] = grouped_negative_sales_month['month_year'].astype(str)

plot_sales(grouped_negative_sales_month, 
    'month_year', 'Sales_EUR', 'Negative Sales by Month', 
    'month_year', 'Total Sales (EUR)')

- At first glance, no evident pattern emerged from the Negative Sales by Month plot. However, the analysis highlights January 2020 as a significant negative month, with a much higher volume of negative sales compared to other months, potentially indicating production errors in a specific product group.

- Excluding this outlier, a subtle pattern can be observed: sales tend to drop noticeably twice a year — once in the first half and once in the second — although the exact months vary.

- This analysis is particularly important because our task involves forecasting product sales based on these values BUT a key part of our role is to uncover patterns that may help prevent future quality issues. If we identify a consistent pattern of returns, we could, with supplementary information, trace the affected product lot and detect potential operational issues that need to be addressed.

In [668]:
negative_sales.loc[:, 'Day'] = negative_sales['DATE'].dt.day
grouped_negative_sales_day = negative_sales.groupby('Day')['Sales_EUR'].sum().reset_index()

grouped_negative_sales_day['Day'] = grouped_negative_sales_day['Day'].astype(str)

plot_sales(grouped_negative_sales_day, 'Day', 'Sales_EUR', 'Negative Sales by Day', 'Day', 'Total Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\3183730899.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



**Observations from the Negative Sales by Day Plot**

The plot reveals a tendency for higher negative sales at the start of the month, which gradually decrease as the month progresses. A secondary peak is also observed around the 20th day. These trends might be explained by:

- Return Logistics: The returns are frequently processed in batches on specific days, especially around the end or beginning of a financial cycle. It is possible that Siemens, as a large manufacturer, may also have centralized financial processes that generate bulk adjustments at the beginning of each month, contributing to the observed spikes.

> Negative sales by Product:

In [669]:
grouped_negative_sales_products= negative_sales.groupby('Mapped_GCK')['Sales_EUR'].sum().reset_index()

plot_sales_by_product(grouped_negative_sales_products, 'Mapped_GCK', 'Sales_EUR', 'Negative Sales by Product', 'Product Group', 'Total Negative Sales (EUR)')

The product groups with the most negative sales, ranked by the highest negative value, are `Group 1 (€2.6M)`, `Group 5 (€2.4M)`, `Group 3 (€1.2M)`, followed by `Group 11 (€520K)`, `Group 6 (€420K)`, and `Group 16 (€290K)`.

This raises some important questions:

- Are these the same products with the highest sales volumes? Is there a proportional relationship between negative sales and positive sales values?

- Considering the frequency of sales (i.e., how many times each product appears every month), are these negative sales values justified? Some products may be significantly more expensive than others, and their returns could lead to greater losses. We will analyze this further:

In [670]:
# Count the number of occurrences (sales) for each product
grouped_sales_count_negative = negative_sales.groupby('Mapped_GCK').size().reset_index(name='Sales_Count')
plot_sales_by_product(grouped_sales_count_negative, 'Mapped_GCK', 'Sales_Count', 'Sales Count by Product', 'Product Group', 'Number of Sales')

- Although Product 1 and Product 5 have the highest values for negative sales, they don’t actually have the highest number of returns, with 51 and 43 returns, respectively. This suggests that these products are likely more expensive, or their returns could involve a large number of components or incur high logistical costs, which would contribute to a higher negative value.

- Product 3 has the most returns in terms of quantity, with 56 sales returned, but its negative sales value is only the third largest, indicating that it may be less expensive. For example, Product 4 has 40 returns and a negative sales value of only €110K, suggesting it is relatively inexpensive.

Analysing the average:


- Product 1: 51 returns, aprox 2.6M negative sales ➡️ 50,980 per return

- Product 5: 43 returns, aprox 2.4M negative sales ➡️ 55,814 per return

- Product 3: 56 returns, aprox 1.2M negative sales ➡️ 21,429 per return

- Product 11: 7 returns, aprox 520K negative sales ➡️ 74,286 per return

- Product 6: 10 returns, aprox 420K negative sales ➡️ 42,000 per return

- Product 16: 6 returns, aprox 290K negative sales ➡️ 48,333 per return

- Product 4: 40 returns, aprox 110K negative sales ➡️ 2,750 per return

As we can see, Product 1, Product 5, and especially Product 11 have significantly higher negative sales per return, which suggests that these products are either more expensive or involve complex components or high logistical costs. Regardless of the reason, these high negative impacts are concerning for the company. We need to investigate whether this issue is localized to specific periods or is more widespread.

In [671]:
negative_sales.loc[:, 'Year'] = negative_sales['DATE'].dt.year

# Group by 'Mapped_GCK' (Product) and 'Year' to calculate the sum of 'Sales_EUR' and count the occurrences
grouped_negative_sales_year = negative_sales.groupby(['Mapped_GCK', 'Year']).agg(
    Sales_Sum=('Sales_EUR', 'sum'),
    Sales_Count=('Sales_EUR', 'size')
).reset_index()

plot_sales_by_product_month(grouped_negative_sales_year, 'Year', 'Sales_Sum', 'Mapped_GCK', 'Negative Sales Trends by Product and Year', 'Year', 'Total Negative Sales (EUR)')

C:\Users\marga\AppData\Local\Temp\ipykernel_15752\4208296278.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



- **2019** was a challenging year for some products. **Product 1** experienced its lowest negative sales (€1.2M), followed by **Product 5**, which also had a significant amount (€770K). Additionally, **Products 11** and **16** saw the majority of their returns in this year. This is somewhat reassuring, as **Product 11** is one of the more expensive items, so it’s good to see that the issues appear to be contained.

- In **2020**, while **Product 1** (€420K) and **Product 5** (€370K) continued to have large returns, **Product 3** saw a notable increase in negative sales (€830K), along with **Product 6** (€333K). This combination of returns had a substantial impact. In **2021**, **Product 5** experienced a significant spike in negative sales (€1M), reinforcing the pattern that **Products 1** and **5** consistently face issues— which is concerning given that these are costly products. In contrast, the other products likely experienced negative sales due to more localized production problems, which may have been minimized over time.

In [672]:
grouped_negative_sales_prod_month = negative_sales.groupby(['month_year', 'Mapped_GCK'])['Sales_EUR'].sum().reset_index()
grouped_negative_sales_prod_month['month_year'] = grouped_negative_sales_prod_month['month_year'].astype(str)

plot_sales_by_product_month(grouped_negative_sales_prod_month, 'month_year', 'Sales_EUR', 'Mapped_GCK', 
                            'Negative Sales by Product and Month', 'month_year', 'Total Sales (EUR)')

The visualization of **Negative Sales by Poduct, by month** gives us an idea of the main drivers for the lower peaks in sales values:
 
- **Mar 2019**: most negative sales for this month were driven by the product group **11**, making it possibly a production error. 
- **Jan 2020**: the month with the lowest value for sales resulted from a variety of product groups having a peak in negative sales, namely group **6**, group **16**, group **3**, group **5** and group **1**.
- **Nov 2020**: while this was a month that looked within normal limits, now its possible to identify that the vast majority of negative sales for November 2020 come from the product group **3**, making it very likely a production error. 
- **Apr 20201**: most negative sales for this month were driven by the product group **5**, with a slight peak in group **1** as well, making it possibly a production error.
- **Sep and Oct 2021**: both this months have a similar behaviour to the one of **Apr 2021**, being the main drivers of negative sales products **5** and **1**.

##### <span style="background-color:#000027; padding:5px; border-radius:5px;">**2.5. Positive Values Analysis**</span> <a id='positive-values'></a> 

In [673]:
positive_sales = df_sales[df_sales['Sales_EUR'] > 0].copy()

In [674]:
grouped_positive_sales_month = positive_sales.groupby('month_year')['Sales_EUR'].sum().reset_index()

grouped_positive_sales_month['month_year'] = grouped_positive_sales_month['month_year'].astype(str)

plot_sales(grouped_positive_sales_month, 'month_year', 'Sales_EUR', 'Positive Sales by Month', 'Year-Month', 'Total Positive Sales (EUR)')

In [675]:
# comparing the percentage change in sales for positive and negative sales
grouped_positive_sales_month['Sales_EUR_pct_change'] = round(grouped_positive_sales_month['Sales_EUR'].pct_change() * 100,2)
grouped_negative_sales_month['Sales_EUR_pct_change'] = round(grouped_negative_sales_month['Sales_EUR'].pct_change() * 100,2)

merged_sales = pd.merge(grouped_positive_sales_month[['month_year', 'Sales_EUR_pct_change']], 
                        grouped_negative_sales_month[['month_year', 'Sales_EUR_pct_change']], 
                        on='month_year', 
                        suffixes=('_positive', '_negative'))

merged_sales['month_year'] = merged_sales['month_year'].astype(str)

table = merged_sales.set_index('month_year').T
table

month_year                     2018-11   2018-12  2019-01  2019-02  2019-03  \
Sales_EUR_pct_change_positive -37.8000  100.1300 -16.3500   5.6400  13.9000   
Sales_EUR_pct_change_negative      NaN 2510.1400 -44.0600 -46.8900 578.0800   

month_year                     2019-04  2019-05  2019-06  2019-07  2019-08  \
Sales_EUR_pct_change_positive  -1.7900   7.8800   1.1300  -2.0100  -4.9700   
Sales_EUR_pct_change_negative -50.4700 -62.4300  51.0100  -5.3200  43.1900   

month_year                     2019-09  2019-10  2019-11  2019-12  2020-01  \
Sales_EUR_pct_change_positive  28.4100 -19.4900  -2.6200   6.3700 -16.0300   
Sales_EUR_pct_change_negative  52.9900  52.7200 -83.3100 554.8600  60.3400   

month_year                     2020-02  2020-03  2020-04  2020-05  2020-06  \
Sales_EUR_pct_change_positive   6.2100  11.4400  -7.9500  -9.1800  17.2400   
Sales_EUR_pct_change_negative -85.3100 119.7000 -62.5000 -88.8800 669.0000   

month_year                     2020-07  2020-08  2020-09    2020-10   2020-11  \
Sales_EUR_pct_change_positive -10.2600  10.8700  17.7900   -23.0600   13.6900   
Sales_EUR_pct_change_negative -51.3600 -50.4600 -98.6700 13242.7000 1281.3800   

month_year                     2020-12  2021-01  2021-02  2021-03  2021-04  \
Sales_EUR_pct_change_positive -14.0200 -41.8500 120.1800 -13.3100   4.9200   
Sales_EUR_pct_change_negative -93.1300 -91.7300 266.8500 359.6200 719.9000   

month_year                     2021-05  2021-06  2021-07  2021-08  2021-09  \
Sales_EUR_pct_change_positive  -7.1700   5.9400  -2.1100  15.0700   3.5400   
Sales_EUR_pct_change_negative -84.6100 -83.3700 154.8500  73.2700 736.5600   

month_year                     2021-10  2021-11  2021-12  2022-01  2022-02  \
Sales_EUR_pct_change_positive  -5.7600  -6.4600  -8.1100   1.8900   0.5300   
Sales_EUR_pct_change_negative -11.9500 -63.8800  18.1800  80.1800 -56.8700   

month_year                     2022-03   2022-04  
Sales_EUR_pct_change_positive  18.5100  -12.1900  
Sales_EUR_pct_change_negative -93.1300 2289.1600

**Nevative vs Positive Sales Values**

- (*Outlier*) **November 2018** was a month of few activity, with the lowest overall positive sales and negative sales close to non.

- **September 2019** is a peak in positive sales, that is followed by a peak in negative sales, in **October 2019** - which might come has a consequence of the high sales in the previous month. 

- **January 2020** is not only the month with highest negative values, but it is also a month categorized by the bellow average positive sales, which is likely evidence of a outside event's affect in the sales behavior.

- (*Outlier*) **January 2021** is a month marked by a abrupt change in sales behavior, it is also a month of few activity, with one of the lowest sales value, by far.

- **September 2021** is a month that peaks in both negative and positive sales values - this kind of situation can be taken as a proportional increase in positive and negative (absolute) sales.

In [676]:
positive_sales['Day'] = positive_sales['DATE'].dt.day
grouped_positive_sales_day = positive_sales.groupby('Day')['Sales_EUR'].sum().reset_index()

grouped_positive_sales_day['Day'] = grouped_positive_sales_day['Day'].astype(str)

plot_sales(grouped_positive_sales_day, 'Day', 'Sales_EUR', 'Positive Sales by Day', 'Day', 'Total Positive Sales (EUR)')

Similar to Negative sales values, the positive sales peaks at the beginning of the month. This aligns with the expected spending behavior coming from enterprises, due to budgeting/cash flow management, payroll and other expenses settlement at the end of the month, between others. 

In [677]:
grouped_positive_sales_product= positive_sales.groupby('Mapped_GCK')['Sales_EUR'].sum().reset_index()

plot_sales_by_product(grouped_positive_sales_product, 'Mapped_GCK', 'Sales_EUR', 'Positive Sales by Product', 'Product Group', 'Total Positive Sales (EUR)')

In [678]:
# Count the number of occurrences (sales) for each product
grouped_sales_count_positive = positive_sales.groupby('Mapped_GCK').size().reset_index(name='Sales_Count')
plot_sales_by_product(grouped_sales_count_positive, 'Mapped_GCK', 'Sales_Count', 'Sales Count by Product', 'Product Group', 'Number of Sales')

In [679]:
grouped_positive_sales_product['Avg_Price_Per_Sale'] = (
    grouped_positive_sales_product['Sales_EUR'] / grouped_sales_count_positive['Sales_Count'])

grouped_positive_sales_product[['Mapped_GCK', 'Avg_Price_Per_Sale']].sort_values(
    by='Avg_Price_Per_Sale', ascending=False)

Mapped_GCK  Avg_Price_Per_Sale
0            1        6089610.6515
1            3        2558118.2809
3            5        1999868.4055
7           11         458069.2968
5            8         170476.0467
4            6         113854.9578
2            4          81144.9202
11          16          59300.3649
8           12          50272.5702
10          14           8047.0732
13          36           7906.2475
9           13           7314.1305
6            9           3391.7151
12          20            967.4875

The fact that **Product 11 has a higher negative impact per return** could suggest that its **return process involves higher infrastructure or logistical costs**. This might not be evident in positive sales because **Products 1, 3, and 5 dominate in average sale price**, meaning their revenue contribution is clearer: 

 - **Product 11 might be more expensive to return**  
   - Unlike **Products 1, 3, and 5, which have the highest positive sales values**, Product 11 does not stand out as much in positive sales but has a **very high cost per return (74K per return)**.  
   - This suggests that **the cost of reversing a sale (e.g., logistics, dismantling, restocking) is exceptionally high**, even if the original sale price is lower than Products 1, 3, or 5.  

- **Product 1 could involve multiple components**  
   - While **Product 1 has the highest average sale price (6M per sale)**, its negative cost per return (~50K per return) is **not as extreme as Product 11**.  
   - This could indicate that **Product 1 is expensive but might involve multiple sub-components**, meaning that a return does not necessarily affect the entire unit, but only parts of it.  
   - The impact of a return might be spread across different elements rather than being concentrated in a single high-cost process.  

- **Returns might not mirror the original sales structure**  
   - In **positive sales**, the highest values are from **Products 1, 3, and 5**, meaning they generate the most revenue.  
   - However, in **negative sales per return**, Product 11 stands out. This discrepancy suggests that **some products may be far more expensive to handle during returns than they are to sell**.  
   - This could be due to **handling complexity, dismantling costs, or high depreciation upon return**.  

In [680]:
# Merge positive sales count with negative sales count
comparison_df = grouped_sales_count_positive.merge(
    grouped_sales_count_negative, on='Mapped_GCK', how='left', suffixes=('_positive', '_negative'))

# Compute percentage of returns per product
comparison_df['Return_Percentage'] = (
    comparison_df['Sales_Count_negative'] / comparison_df['Sales_Count_positive']) * 100

# Merge with average price per sale
comparison_df = comparison_df.merge(
    grouped_positive_sales_product[['Mapped_GCK', 'Avg_Price_Per_Sale']],on='Mapped_GCK',how='left')
comparison_df[['Mapped_GCK', 'Sales_Count_positive', 'Sales_Count_negative', 'Return_Percentage']]


Mapped_GCK  Sales_Count_positive  Sales_Count_negative  Return_Percentage
0            1                   254                    51            20.0787
1            3                   209                    56            26.7943
2            4                   194                    40            20.6186
3            5                   216                    43            19.9074
4            6                   185                    10             5.4054
5            8                   205                    19             9.2683
6            9                    94                    11            11.7021
7           11                   145                     7             4.8276
8           12                   165                    10             6.0606
9           13                   138                     3             2.1739
10          14                    63                     7            11.1111
11          16                   315                     6             1.9048
12          20                    81                     5             6.1728
13          36                   128                     8             6.2500

- Product 3 has the highest number of returns (56) and the highest return rate (26.8%), suggesting either a quality issue or a mismatch in customer expectations. However, we observed that this was mostly localized in November 2020, so the underlying issue may have already been resolved.

- Product 1 and Product 5 are high-value products, and a 20% return rate is financially significant. However, their return rates remain bad across different years.

- Product 4 has a low per-sale value but still a high return percentage, suggesting that it might be a lower-cost product that is frequently returned.

In [681]:
grouped_positive_sales_prod_month = positive_sales.groupby(['month_year', 'Mapped_GCK'])['Sales_EUR'].sum().reset_index()
grouped_positive_sales_prod_month['month_year'] = grouped_positive_sales_prod_month['month_year'].astype(str)

In [682]:
# linear trend line
plot_sales_trends(grouped_positive_sales_prod_month, trend_type='linear')

In [683]:
plot_sales_trends(grouped_positive_sales_prod_month, trend_type='rolling')

- Product 1: No clear trend.

- Product 3: Highly volatile, but overall has an increasing trend - can be mearly due to inflation.

- Product 4: Irregular pattern with no strong trend, suggesting it may be stationary.

- Product 5: Irregular pattern with no strong trend, suggesting it may be stationary.

- Product 6: decreasing trend with fluctuations.

- Product 8: Increasing trend, with some spikes in September, decline in October. Likely seasonal.

- Product 9: Irregular pattern with no strong trend, suggesting it may be stationary.

- Product 11: Spikes in September, but no consistent trend. Likely seasonal and stationary.

- Product 12: Growing trend, with fluctuations. Not strongly seasonal.

- Product 13: No clear trend, likely stationary.

- Product 14: No clear trend, likely stationary.

- Product 16: Downward trend, possibly stationary.

- Product 20: No clear trend, likely stationary.

- Product 36: No clear trend, likely stationary.

To get a better understanding of the impact of Negative Sales' values in Positive Sales:

In [ ]:
# Negative sales to absolute values
negative_sales.loc[:, 'Sales_EUR'] = negative_sales['Sales_EUR'].abs()
# Group by 'month_year' and 'Mapped_GCK' to calculate the sum of 'Sales_EUR'
grouped_neg_abs_sales_prod_month = negative_sales.groupby(['month_year', 'Mapped_GCK'])['Sales_EUR'].sum().reset_index()
grouped_neg_abs_sales_prod_month['month_year'] = grouped_neg_abs_sales_prod_month['month_year'].astype(str)

In [ ]:
# Replacing missing month_year with 0, for categories that did not have negative sales in that month
# Define the complete date range we want to include in the analysis
full_date_range = pd.date_range(start='2018-10-01', end='2022-04-01', freq='MS').strftime('%Y-%m').tolist()

# Get unique categories of 'Mapped_GCK'
unique_categories = grouped_neg_abs_sales_prod_month['Mapped_GCK'].unique()

# Create a full dataframe with all combinations of 'month_year' and 'Mapped_GCK'
full_index = pd.MultiIndex.from_product([full_date_range, unique_categories], names=['month_year', 'Mapped_GCK'])
full_df = pd.DataFrame(index=full_index).reset_index()

# Merge the full dataframe with the original one
grouped_neg_abs_sales_prod_month = full_df.merge(grouped_neg_abs_sales_prod_month, on=['month_year', 'Mapped_GCK'], how='left')

# Fill missing values with 0
grouped_neg_abs_sales_prod_month.fillna(0, inplace=True)

In [685]:
# Convert 'month_year' to datetime format in both dataframes
grouped_positive_sales_prod_month['month_year'] = pd.to_datetime(grouped_positive_sales_prod_month['month_year'])
# Remove the day from 'month_year' in grouped_positive_sales_prod_month
grouped_positive_sales_prod_month['month_year'] = grouped_positive_sales_prod_month['month_year'].dt.to_period('M')
grouped_positive_sales_prod_month['month_year'] = grouped_positive_sales_prod_month['month_year'].astype(str)

In [690]:
# Merge datasets on 'month_year' and 'Mapped_GCK'
merged_sales = pd.merge(
    grouped_positive_sales_prod_month[['month_year', 'Mapped_GCK', 'Sales_EUR']],
    grouped_neg_abs_sales_prod_month[['month_year', 'Mapped_GCK', 'Sales_EUR']],
    on=['month_year', 'Mapped_GCK'],
    suffixes=('_positive', '_negative')
)

# Calculate the difference between positive and negative sales
merged_sales['Sales_EUR_diff'] = merged_sales['Sales_EUR_positive'] - merged_sales['Sales_EUR_negative']

# Get unique product categories
unique_products = merged_sales['Mapped_GCK'].unique()

# Create subplot layout (1 row, multiple columns)
fig = make_subplots(
    rows=1, cols=len(unique_products),
    subplot_titles=[f'Product {p}' for p in unique_products])

# Loop through products and add subplots
for i, product in enumerate(unique_products):
    product_data = merged_sales[merged_sales['Mapped_GCK'] == product]

    # Add single line for sales difference
    fig.add_trace(go.Scatter(
        x=product_data['month_year'], 
        y=product_data['Sales_EUR_diff'], 
        mode='lines', 
        name=f'Sales Difference - {product}', 
        line=dict(color='blue')
    ), row=1, col=i+1)

    fig.add_trace(go.Scatter(
        x=product_data['month_year'], 
        y=[0] * len(product_data),
        mode='lines',
        name=f'Zero Line - {product}',
        line=dict(color='black', dash='dash')
    ), row=1, col=i+1)

# Update layout for better visualization
fig.update_layout(
    title='Impact of Negative values in Sales by Product',
    height=500, width=300 * len(unique_products),  # Adjust width dynamically
    showlegend=True)

# Show the figure
fig.show()

Products that have Negative Sales'values for certain months:
- Product 6 (Jan 2020)
- Product 9 (Nov 2020)
- Product 14 (Sep 2021)

**Product 1**: Negative values no sinificant effect on sales. Comparing this plot with the positive sales behavior, there is no visible change.

**Product 3**: Negative values no sinificant effect on sales. Comparing this plot with the positive sales behavior, there is no visible change.

**Product 4**: Negative values no sinificant effect on sales. Comparing this plot with the positive sales behavior, there is no visible change.

**Product 5**: Negative values no sinificant effect on sales. Comparing this plot with the positive sales behavior, there is no visible change.

**Product 6**: While it seems that most months sales values are not hugely affected by the negative values, there is an exception to the rule - **January 2020**. The sales in this month were not the highest, neither the lowest, which confirms the irregular behavior of negative values in this specific month. 